ansible_base/activitystream/models/entry.py

```diff
@@ -0,0 +1,83 @@
+from functools import partial
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.signals import m2m_changed, post_save, pre_delete, pre_save
+from django.utils.translation import gettext_lazy as _
+
+from ansible_base.activitystream.signals import activitystream_create, activitystream_delete, activitystream_update, activitystream_m2m_changed
+from ansible_base.lib.abstract_models import CommonModel
+
+
+class Entry(CommonModel):
+    """
+    An activity stream entry.
+
+    This is keyed on a generic object_id and content_type, which allows for
+    a wide variety of objects to be used in the activity stream.
+    """
+
+    allow_anonymous_user_save = True
+
+    class Meta:
+        verbose_name_plural = _('Entries')
+
+    OPERATION_CHOICES = [
+        ('create', _('Entity created')),
+        ('update', _("Entity updated")),
+        ('delete', _("Entity deleted")),
+        ('associate', _("Entity was associated with another entity")),
+        ('disassociate', _("Entity was disassociated with another entity")),
+    ]
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING)
+    object_id = models.TextField(null=True, blank=True)
+    content_object = GenericForeignKey('content_type', 'object_id')
+    operation = models.CharField(max_length=12, choices=OPERATION_CHOICES)
+    changes = models.JSONField(null=True, blank=True)
+
+    # This is used for m2m (dis)associations
+    related_content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING, null=True, blank=True, related_name='related_content_type')
+    related_object_id = models.TextField(null=True, blank=True)
+    related_content_object = GenericForeignKey('related_content_type', 'related_object_id')
```

Comment:
A question I am asking myself on the RBAC side - for any GFK, should we link to the `Resource` entry for the object? I'm talking about a nullable ForeignKey (to accommodate models it is not enabled for). I'm not selling this as the replacement for the GFK, but if we have any asks for showing the resource `ansible_id` or anything, it become _dead simple_ with this method, and the other approaches we have available work but are much more questionable.

Reply:
> I'm not selling this as the replacement for the GFK

So you are envisioning an optional/nullable FK field _in addition_ to the GFK, that points to the `Resource` entry?

So, I think I'm okay with that, but what implications does this come with? If an app doesn't enable `dab_resource_registry` in `INSTALLED_APPS` - will referencing the model in an FK still work?

And then - assuming that does work fine - where does the value come from? Like, how do the signals know what the FK should point to? I suppose we could have a class var that models can set and points to the name of the field that contains it or something?

Reply:
> So, I think I'm okay with that, but what implications does this come with? If an app doesn't enable `dab_resource_registry` in `INSTALLED_APPS` - will referencing the model in an FK still work?

This is _exactly_ what I'm struggling with myself. It would make the resource registry a hard dependency of the app doing this. Migrations would have to list it as a dependency so that the Resource model exists. Up until this point the RBAC lib is a general Django RBAC list, the activity stream is a general Django activity stream, and I'd hate to start crippling that.

> where does the value come from?

The UI needs the ansible_id in order to navigate related objects between multiple systems. This mostly impacts shared resources like organizations. For a specific case... you can see an organization's activity stream which gives CRUD for the org itself, but maybe that links to the activity stream for AWX resources in the organization or something like that.

> Like, how do the signals know what the FK should point to?

The `Resource` model, itself, has a GFK to the object, so right before creating the activity stream entry, it's simple to just look up that id and add it to the creation data.

Reply:
Doing `Resource.objects.get` incurs an extra query for each serializer wheras a related `self.resource` can be a table JOIN. This would also let clients GET `/activity_stream/?resource__resource_id=123-232-232`

Reply:
@AlanCoding Where did you land on this for RBAC? Do you still think we should have a resource field here?

Reply:
I don't know if there's usable cross-over to RBAC or resource registry right now. For one, no system using DAB RBAC is planned to use this. Then, we don't track a resource object for items other than users/teams/organizations. While integrating the resource could be used for those... that doesn't seem helpful in any way.